ffi/diplomat/src/bidi.rs

```diff
@@ -0,0 +1,156 @@
+            if let Result::Ok(bidi) = maps::get_bidi_class(&provider) {
+                Ok(Box::new(ICU4XBidi(bidi))).into()
+            } else {
+                Err(()).into()
+            }
+        }
+
+        #[diplomat::rust_link(unicode_bidi::BidiInfo::new_with_data_source, FnInStruct)]
+        pub fn segment_text<'text>(&self, text: &'text str) -> Box<ICU4XBidiInfo<'text>> {
+            let data = self.0.get();
+            let adapter = BidiClassAdapter::new(&data.code_point_trie);
+
+            Box::new(ICU4XBidiInfo(BidiInfo::new_with_data_source(
+                &adapter, text, None,
+            )))
+        }
+    }
+
+    #[diplomat::rust_link(unicode_bidi::BidiInfo, Struct)]
+    #[diplomat::opaque]
+    pub struct ICU4XBidiInfo<'text>(pub BidiInfo<'text>);
+
+    impl<'text> ICU4XBidiInfo<'text> {
+        pub fn paragraph_count(&self) -> usize {
+            self.0.paragraphs.len()
+        }
+
+        // The lifetime here should be 'info, not 'text, but we're currently hitting
+        // https://github.com/rust-diplomat/diplomat/issues/142. Fortunately, covariance comes
+        // to our rescue, using 'text here is sufficient to convince the compiler that this is fine
+        pub fn paragraph_at(&'text self, n: usize) -> Option<Box<ICU4XBidiParagraph<'text>>> {
+            self.0
+                .paragraphs
+                .get(n)
+                .map(|p| Box::new(ICU4XBidiParagraph(Paragraph::new(&self.0, p))))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct ICU4XBidiParagraph<'info>(pub Paragraph<'info, 'info>);
+
+    impl<'info> ICU4XBidiParagraph<'info> {
+        #[diplomat::rust_link(unicode_bidi::Paragraph::level_at, FnInStruct)]
+        pub fn direction(&self) -> ICU4XBidiDirection {
+            self.0.direction().into()
+        }
+
+        #[diplomat::rust_link(unicode_bidi::ParagraphInfo::len, FnInStruct)]
+        pub fn size(&self) -> usize {
+            self.0.para.len()
+        }
+
+        /// The start index of this paragraph within the source text
+        pub fn range_start(&self) -> usize {
+            self.0.para.range.start
+        }
+
+        /// The end index of this paragraph within the source text
+        pub fn range_end(&self) -> usize {
+            self.0.para.range.end
+        }
+
+        #[diplomat::rust_link(unicode_bidi::Paragraph::level_at, FnInStruct)]
+        pub fn reorder_line(&self, range_start: usize, range_end: usize, out: &mut DiplomatWriteable) -> DiplomatResult<(), ()> {
+            if range_start < self.range_start() || range_end > self.range_end() {
+                return Err(()).into()
+            }
+
+            let info = self.0.info;
+            let para = self.0.para;
+
+            let reordered = info.reorder_line(para, range_start..range_end);
+
+            out.write_str(&reordered).map_err(|_| ()).into()
+        }
+
+        /// Get the BIDI level. This integer is conceptually a `unicode_bidi::Level`,
```

Comment:
Diplomat doesn't support type aliases with methods or non opaque structs with methods, so oru current choices here are to just return an integer and document its behavior, or return an opaque type

Reply:
For FFI purposes I think it's fine and expected to unwrap newtypes into their inner type.